### ISSUE TITLE
Gem install fails with SSL certificate connection error

### ISSUE THREAD

**Issue by PavelEmelyanov** [rubygems/rubygems#1736]
Hello, I'm having a problem. Today, when I try to execute "gem install ..." or "gem update" I have the following error:

```
C:\WINDOWS\system32>gem update --system
ERROR:  While executing gem ... (Gem::RemoteFetcher::FetchError)
    SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certif
icate verify failed (https://api.rubygems.org/specs.4.8.gz)
```

or

```
C:\WINDOWS\system32>gem install sass compass
ERROR:  Could not find a valid gem 'sass' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect retur
ned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (
https://api.rubygems.org/specs.4.8.gz)
ERROR:  Could not find a valid gem 'compass' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect retur
ned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (
https://api.rubygems.org/specs.4.8.gz)
```

I've updated ruby, but is doesn't help.

This issue is related to:
  - Network problems
  - The command line `gem`

Here are my current environment details:
gem 2.4.5.1
ruby 2.2.5p319 (2016-04-26 revision 54774) [x64-mingw32]

Thanks

**Comment by K-4U** [comment]
For puphpet users: http://pastebin.com/cjW6GtPW

**Comment by segiddins** [comment]
You may need to update your Ruby to use a newer version of OpenSLL

**Comment by PavelEmelyanov** [comment]
Well, I found this on http://rubyinstaller.org/downloads/

> If you don’t know what version to install and you’re getting started with Ruby, we recommend you use Ruby 2.2.X installers.

So I updated my ruby to 2.2.5, but it didn't help me. Should I update ruby to 2.3.1?

**Comment by ptrikutam** [comment]
I had a similar issue:

```
$ gem install bundler
ERROR:  Could not find a valid gem 'bundler' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=error: certificate verify failed (https://api.rubygems.org/specs.4.8.gz)
```

I tried following a number of suggestions provided in https://github.com/rubygems/rubygems/issues/1050, including the solution here: https://gist.github.com/luislavena/f064211759ee0f806c88#installing-using-update-packages-new


I'm running:
rvm  1.27.0
ruby ruby 2.3.1p112 (via RVM)
gem  2.6.7

The solution that worked for me is updating my OS X certs with brew [(took the suggestions from here)](http://railsapps.github.io/openssl-certificate-verify-failed.html) :

```
brew update
brew install openssl
brew tap raggi/ale
brew install openssl-osx-ca
```

Then, doing a `gem update --system` and `gem install bundler` worked fine. Hope this helps.

**Comment by ckenst** [comment]
@ptrikutam thanks for the link to [OpenSSL Errors and Rails – Certificate Verify Failed](http://railsapps.github.io/openssl-certificate-verify-failed.html). Following the instructions for the section Use RVM to Fix SSL Certificates worked for me.

**Comment by ptrikutam** [comment]
@ckenst Glat to hear it. For what it's worth, using RVM to fix the certs didn't work for me, I had to use brew.

**Comment by NarineM** [comment]
I am getting similar error with gem install mysql2:

WARNING:  Unable to pull data from 'https://rubygems.org/': SSL_connect returned                                     =1 errno=0 state=SSLv3 read server certificate B: certificate verify failed 
(https://api.rubygems.org/specs.4.8.gz)

Is this a temp issue and will go away? I am using 2.4.5 version of ruby gem. I was able to make gem download the mysql2 using http by running this command:
gem sources -a http://rubygems.org

But I do not think this could be used as long term solution.

**Comment by zymzxq** [comment]
Same error :(

**Comment by gorenm** [comment]
Same here. on Windows 10. Ruby 2.2.4.
Happens when trying to install bundler.

**Comment by Azolo** [comment]
So, the reason why this is happening is because rubygems.org updated their SSL certificate and older versions don't have the updated certificate bundled.

If you're using a traditional OpenSSL distribution then updating it would likely pull in new certificates that include the new rubygems certificate.

If you're on Windows(where the certifications aren't included) or don't have or can't update system certifications for whatever reason, then you will have to update `rubygems` through other means. See https://gist.github.com/luislavena/f064211759ee0f806c88#installing-using-update-packages-new, but get a more recent updated package.

Unfortunately, there really isn't any other way to solve this issue other than updating your system certificates (impossible for typical Windows users), obtaining the new certificate manually, or through a manually updating `rubygems`.

**Comment by PavelEmelyanov** [comment]
Thank you, manual updating of rubygem helped me.

**Comment by wwwbeck** [comment]
change to use http protocol over https.
1. remove source: gem --remove https://rubygems.org
2. add source: gem --add http://rubygems.org

**Comment by djberg96** [comment]
To followup @wwwbeck's comment, I noticed that rubygems could be persnickety about trailing slashes. If you do `gem sources -v`, you might see "https://rubygems.org/", in which case you'll need to do `gem sources -r https://rubygems.org/`.

**Comment by deveth0** [comment]
I noticed this problem with a fresh installation on windows using the ruby-installer.
Another (imho better workaround) is to replace the https source temporary, run `gem update --system` and then switch back to https.
````
    gem sources --remove https://rubygems.org/
    gem sources --add http://rubygems.org
    gem update --system
    gem sources --remove http://rubygems.org
    gem sources --add https://rubygems.org
````

**Comment by vidavidorra** [comment]
@deveth0 Worked for me (on Windows 10) using the latest installer for 2.3.1 (x64) from [link](http://rubyinstaller.org/downloads). And running the commands you suggested.
Thanks a lot.

**Comment by sangar** [comment]
Thanks @deveth0. This worked for me but I had to manually edit `~/.gemrc` :sources: [], when `gem sources --add http://rubygems.org`  gave me 
```
ERROR:  While executing gem ... (Gem::OperationNotSupportedError)
    Not connected to a tty and no default specified
```

**Comment by manibk31** [comment]
The bus is with connecation establishment over https. Use http instead.
gem source --add http://rubygems.org

This worked for me. (Windows)

**Comment by medghaim** [comment]
http is unsecure and can leave you vulnerable to a MITM attack

**Comment by manibk31** [comment]
Yes using http over https exposes you to MITM attacks.for local development and testing purposes http can be given a shot. In most of the cloud hosting dependencies installation is taken care by the vendor.

**Comment by ferhoba** [comment]
Hello shiny happy people,

I have win 10 - 64x

Manual set up instructions on this link solved issue. http://guides.rubygems.org/ssl-certificate-update/#manual-solution-to-ssl-issue

I hope this could help someone

Greetings.

**Comment by tristantao** [comment]
I highly recommend people reading this thread as well: https://github.com/rubygems/rubygems/issues/1758 It contained the only solution that worked for me (and I tried many).

**Comment by yangshun** [comment]
I simply had to change to ruby-2.3.0 and it installed fine for me.

**Comment by riasol** [comment]
:sources: ["http://rubygems.org"] in .gemrc solved on win7. thnx.

**Comment by cannahum** [comment]
Updating ruby version (to 2.4.1) worked for me too...

**Comment by niefermar** [comment]
gem install rhc --source http://rubygems.org

**Comment by her** [comment]
I think it was an openssl error for me, something about downloading openssl first and then maybe updates breaking somewhere along the way after `brew install [pkg]` multiple times. I removed openssl, reinstalled, that seemed to have fixed it.

**Comment by ashutoshag** [comment]
The certificate changed. Just run the following command, if you are using OSX: 

rvm osx-ssl-certs update all

**Comment by afroguy16** [comment]
Guys read ferhoba. Trust me, it works like charm. 

![image](https://user-images.githubusercontent.com/15367904/29822658-1c69149a-8cc4-11e7-8b81-dbd3ef6148aa.png)

**Comment by BharatBabbar28** [comment]
Hi, you should exclude your system from ssl or https scanning first then try 
It will work definitely otherwise let me know
bharatbabbar28@gmail.com

**Comment by falilou13** [comment]
C:\Windows\system32>gem install rails --no-ri --no-rdoc

C:\Windows\system32>gem install rails --no-ri --no-rdoc
Fetching: nokogiri-1.8.2-x64-mingw32.gem (  3%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem (  4%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem (  3%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem ( 17%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem (  8%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem ( 15%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem (  8%)Fetching: nokogiri-1.8.2Fetching: nokogiri-1.8.2-x64-mingw32.gem ( 10%)ERROR:  While executing gem ... (Gem::RemoteFetcher::FetchError)
    too many connection resets (https://api.rubygems.org/gems/nokogiri-1.8.2-x64-mingw32.gem)

can't install rails on win 10 x-64 please need help

**Comment by ollej** [comment]
I had this issue, and it turned out the rubygems.org certificate was newer than the time on my vagrant machine. Setting the correct time in the vagrant fixed the issue for me.

**Comment by p4ndepravity** [comment]
I didn't want to update rubygems or any other gem because i'm following a tutorial. I followed the link posted by ferhoba. All i did was download the new cert and save it in my ssl_certs folder and i was able to install everything fine.

**Comment by xpagesbeast** [comment]
Used the installer Ruby+Devkit 2.4.X https://github.com/oneclick/rubyinstaller2/releases/download/rubyinstaller-2.4.4-1/rubyinstaller-devkit-2.4.4-1-x64.exe and tried again with success.

**Comment by fayaj1** [comment]
ERROR:  Could not find a valid gem 'cocoapods' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - SSL_connect returned=1 errno=0 state=SSLv2/v3 read server hello A: tlsv1 alert protocol version (https://rubygems.org/latest_specs.4.8.gz)


How to Fix it??

**Comment by btyss-ui** [comment]
I was able to solve my problem by just installing my environment through Linux Ubuntu subsystem (I'm on Windows)

**Comment by gptshubham595** [comment]
Download the cacert.pem file from http://curl.haxx.se/ca/cacert.pem. Save this file to C:\RailsInstaller\cacert.pem.

Now make ruby aware of your certificate authority bundle by setting SSL_CERT_FILE. To set this in your current command prompt session, type:

set SSL_CERT_FILE=C:\RailsInstaller\cacert.pem

**Comment by mze3e** [comment]
> The certificate changed. Just run the following command, if you are using OSX:
> 
> rvm osx-ssl-certs update all

This worked

**Comment by mohitgpt24** [comment]
When I try to bundle I am getting 

```
There is a chance you are experiencing a man-in-the-middle attack, but most
likely your system doesn't have the CA certificates needed for verification. For
information about OpenSSL certificates.
```
None of the above solutions is working. I am using macOS BigSur.

**Comment by kannannadarajan** [comment]
> > The certificate changed. Just run the following command, if you are using OSX:
> > rvm osx-ssl-certs update all
> 
> This worked

This worked me too in MacOs BigSur

Thank you all!!

**Comment by Palpandi6295** [comment]
> Hi Team, one of my customers git version upgraded from 2.39 to 2.44 and they started facing the SSL error. The traffic is going through the proxy and they have imported the proxy certificate in to the folder with crt and pem format same issue If they downgrade back to 2.39 then the issue is getting resolved or if they do SSL bypass on 2.44 on the proxy also resolved the issue.
> 
> PS C:\CE REPO\gidp-ce-casemanagement-sfvc> git pull fatal: unable to access 'https://bitbucket.org/nttdataamericas/gidp-ce-casemanagement-sfvc.git/': Failure when receiving data from the peer
> 
> <img alt="image" width="678" src="https://private-user-images.githubusercontent.com/162577134/312004065-cebcd88d-0ce5-41e4-b4d7-aeb6d6ce4bcf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMTAzOTEsIm5iZiI6MTcxMTExMDA5MSwicGF0aCI6Ii8xNjI1NzcxMzQvMzEyMDA0MDY1LWNlYmNkODhkLTBjZTUtNDFlNC1iNGQ3LWFlYjZkNmNlNGJjZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxMjIxMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYTdkY2Q4NWE0NDg4NWMzMDc3MjA1ZTVjNjZhMmMyZmQxNzg0OTc4OWRlMjZmZGU3ZmNmODUzODIzMjMwY2VkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.I_Tdm7HGVLuJuyWjlPJSfnFWXt7v2-NQTPoIEtoJIZw">
> Can some one please check and help on with priority to proceed further

Hi Experts,
Do we have any update on this above query?
Since it is affecting the customer environment, need your help on this to proceed further

Regards,
Palpandi R
